### ISSUE TITLE
successful_active_health_check_uneject_host can cause not wanted outlier uneject

### ISSUE THREAD

**Issue by tkovacs-2** [envoyproxy/envoy#32442]
Title: `successful_active_health_check_uneject_host` can cause not wanted outlier uneject

Description:
This callback function:
https://github.com/envoyproxy/envoy/blob/b5ca88acee3453c9459474b8f22215796eff4dde/source/common/upstream/outlier_detection_impl.cc#L315
doesn't care about the result of the health check or the health transition parameter, only the health flags.

Consider the following scenario:
1. The peer host becomes not available.
2. `FAILED_ACTIVE_HC` health flag is still not set because there was no or not enough failed health check so far.
3. Outlier detection ejects the host.
4. If the next health check still doesn't set `FAILED_ACTIVE_HC`, it will uneject the host regardless of whether the check result was success or failure.

You can see this in a real health check and outlier detection log:

Host `127.0.0.1:40002` is down, first failed health check:
```
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:01.267Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
```

Outlier detection ejects the host:
```
{"type":"CONSECUTIVE_5XX","timestamp":"2024-02-13T16:15:03.083Z","cluster_name":"cluster_0","upstream_url":"127.0.0.1:40002","action":"EJECT","num_ejections":1,"enforced":true,"eject_consecutive_event":{}}
```

Second failed health check. Callback unejects the host which makes its state healthy, but this is obviousy wrong:
```
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:06.268Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"type":"CONSECUTIVE_5XX","timestamp":"2024-02-13T16:15:06.268Z","secs_since_last_action":"3","cluster_name":"cluster_0","upstream_url":"127.0.0.1:40002","action":"UNEJECT","num_ejections":1,"enforced":false}
```

Outlier detectiont ejects the host again:
```
{"type":"CONSECUTIVE_5XX","timestamp":"2024-02-13T16:15:09.084Z","secs_since_last_action":"2","cluster_name":"cluster_0","upstream_url":"127.0.0.1:40002","action":"EJECT","num_ejections":2,"enforced":true,"eject_consecutive_event":{}}
```

Third failed health check finally sets `FAILED_ACTIVE_HC`, callback does nothing just like at the further health checks.
```
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","eject_unhealthy_event":{"failure_type":"NETWORK"},"timestamp":"2024-02-13T16:15:11.269Z","locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:11.269Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:16.270Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:21.271Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:26.273Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
{"health_checker_type":"THRIFT","host":{"socket_address":{"protocol":"TCP","address":"127.0.0.1","port_value":40002,"resolver_name":"","ipv4_compat":false}},"cluster_name":"cluster_0","timestamp":"2024-02-13T16:15:31.274Z","health_check_failure_event":{"failure_type":"NETWORK","first_check":false},"locality":{"region":"","zone":"","sub_zone":""}}
```

The same has been already mentioned in #28303, but closed without any real outcome.

Since uneject can happen even at failed health check I doubt this was a planned behaviour for `successful_active_health_check_uneject_host`.
For fix `&& transition == HealthTransition::Changed` should be added to the condition in the callback to 
make sure we uneject the host only at that health check which just cleared the `FAILED_ACTIVE_HC` flag.

Config:
```
...
cluster_manager:
  outlier_detection:
    event_log_path: /tmp/thrift_health.log
static_resources:
  listeners:
  - name: listener_0
    address:
      socket_address:
        protocol: TCP
        address: 127.0.0.1
        port_value: 50051
    filter_chains:
    - filters:
      - name: envoy.filters.network.thrift_proxy
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.filters.network.thrift_proxy.v3.ThriftProxy
          stat_prefix: ingress_service
          protocol: BINARY
          transport: FRAMED
          payload_passthrough: true
          route_config:
            ...
          thrift_filters:
          - name: envoy.filters.thrift.router
            typed_config:
              "@type": type.googleapis.com/envoy.extensions.filters.network.thrift_proxy.router.v3.Router
              close_downstream_on_upstream_error: false
  clusters:
  - name: cluster_0
    ...
    common_lb_config:
      healthy_panic_threshold:
        value: 0.0
    health_checks:
    - timeout: "5s"
      interval: "5s"
      no_traffic_interval: "10s"
      unhealthy_threshold: 3
      healthy_threshold: 1
      event_log_path: /tmp/thrift_health.log
      always_log_health_check_failures: true
      custom_health_check:
        name: envoy.health_checkers.thrift
        typed_config:
          "@type": type.googleapis.com/envoy.extensions.health_checkers.thrift.v3.Thrift
          method_name: ping
          transport: FRAMED
          protocol: BINARY
    outlier_detection:
      consecutive_5xx: 2
      interval: "10s"
      base_ejection_time: "30s"
      max_ejection_time: "30s"
      max_ejection_percent: 100
```

**Comment by kyessenov** [comment]
CC @wbpcode for outlier detection expertise

**Comment by ramaraochavali** [comment]
> && transition == HealthTransition::Changed 

this was the original implementation but later removed based on this [comment](https://github.com/envoyproxy/envoy/pull/24331#discussion_r1058551719).  I will take a look on Monday to see if it makes sense to add that check back.
@wbpcode  for any other suggestions

**Comment by tkovacs-2** [comment]
Yes, it depends on what we mean under the `successful_active_health_check_` part.
A single successful check or `healthy_threshold` long series of successful checks.

Since the callback is called after both success and failed checks some additional condition is needed there.
For me the second interpretation seems more reasonable, that is why I suggested the transition.
The first is also ok however little weird: If `healthy_threshold` is bigger than 1 nobody expect any action for a single success hc.
Also today a callback doesn't get information about the result of the single health check.

**Comment by ramaraochavali** [comment]
> Yes, it depends on what we mean under the successful_active_health_check_ part.
A single successful check or healthy_threshold long series of successful checks.

A single successful check is what was implemented because if we rely on healthy threshold (>1 case), the outlier detection might actually "uneject" because its ejection period has passed - so this flag might not be of that use. This flag was introduced to get the benefit of a successful check before the ejection time so that we do not have to wait longer.

**Comment by tkovacs-2** [comment]
Yes, if healthy threshold is big, ejection period is small, then this flag wouldn't have effect. But why is that problem?
But ok, let's stay at the original concept. How the callback can get the result of the current single health check?

**Comment by NersesAM** [comment]
it seems the changelog entry for this bug fix was lost during merge conflict resolution. Can it be corrected? Was it backported to 1.29.x?

**Comment by phlax** [comment]
hi @NersesAM that is just in the `latest` docs - i was about to update this now anyway - the docs for the supported  branches should be correct already i think
